### PR TITLE
Integration with the SDK (MVP cont.)

### PR DESCRIPTION
**Pull request by digorithm**
**EDIT(iqdecay):** I have taken over this PR, and would like to get this merged "as is", with the refactoring being done in #11 and some more work being done aside.

demo

Continuing the features documented here https://github.com/FuelLabs/forc-wallet/issues/2. 

This PR integrates the forc wallet plugin with the SDK, adds wallet encryption and storage, wallet `list`ing. 

```
USAGE:
    forc-wallet [OPTIONS] <SUBCOMMAND>

OPTIONS:
    -h, --help            Print help information
    -o, --format <FMT>    [default: json] [possible values: json, toml]
    -V, --version         Print version information

SUBCOMMANDS:
    help      Print this message or the help of the given subcommand(s)
    import    Import a wallet from mnemonic phrase
    list      Lists all wallets stored in ~/.fuel/wallets/
    new       Randomly generate a new wallet. By default, wallets are stored in ~/.fuel/wallets/
```

Wallet creation now requires a password for the encryption process. Password input is protected and the outputted wallet is outputted in a safe way that can't be stored in the terminal's history. 

Wallets are stored in `~/.fuel/wallets`. To open these wallets you need the password you used to create them.

Creating the wallet:

```
./target/debug/forc-wallet new
Please enter a password to encrypt this private key:
Please confirm your password:
```

Then the safe output:

```
JSON Wallet uuid: 3ef27797-fd47-45da-b9a1-6d4f7b1f5164

Wallet public address: 3069ea2d1cca428fbe91137943dfaa5f193695730ef62d1f940fee7ffa465326

Wallet mnemonic phrase: space agree swing obvious hover chimney liquid tonight never pony electric useless

### Do not share or lose this mnemonic phrase! Press any key to complete. ###
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,13 @@ anyhow = "1.0"
 clap = { version = "3.1", features = ["derive"] }
 fuel-crypto = "0.4"
 fuel-types = "0.3"
+fuels = { git = "https://github.com/fuellabs/fuels-rs", branch = "master" }
```
Comment (digorithm) [inline Cargo.toml:17]: Temporarily tagging `master` branch because the features needed by this PR haven't been released yet.

Comment (adlerjohn) [reply Cargo.toml:17]: In the interest of not having things break, can we hold off on this PR until the feature is released?

Comment (digorithm) [reply Cargo.toml:17]: Yup, that's the plan!

**Comment by adlerjohn** [comment]
Converting to draft until unblocked by new `fuels` release.

**Comment by iqdecay** [comment]
Only pushing version change in this PR, will add tests and other features in other more focused PRs.

File: Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,13 @@ anyhow = "1.0"
 clap = { version = "3.1", features = ["derive"] }
 fuel-crypto = "0.4"
```
Comment (kayagokalp) [inline Cargo.toml:15]: Should we also use the latest `fuel-crypto (v0.5)` and `fuel-types (v0.5.1)` (I guess they wouldn't break as they seem to be used in the latest version of [`fuels-rs`](https://github.com/FuelLabs/fuels-rs/blob/bca50740445a4dc3f09386cd2ed546de06f12ee0/packages/fuels-core/Cargo.toml#L15)).

Comment (iqdecay) [reply Cargo.toml:15]: Definitely! good catch

**Review by iqdecay — APPROVED** [review]
Shameless self-approval :)

**Review by kayagokalp — DISMISSED** [review]
LGTM :shipit:

**Review by kayagokalp — DISMISSED** [review]
lgtm

File: src/main.rs [hunk 3]
```diff
@@ -36,31 +46,153 @@ enum OutputFormat {
...
+                .collect();
+            let index = split.first().unwrap().parse::<usize>().unwrap();
+            if index > highest_index {
+                highest_index = index;
+            }
+        }
+    }
+
+    highest_index + 1
 }
```
Comment (segfault-magnet) [inline src/main.rs:150]: A functional alternative, for the `.map` lovers :D 
```suggestion
// Walks through ~/.fuel/wallets/ and returns the next index based
// on the number of wallets in the vault.
fn get_next_wallet_index() -> usize {
    let path = home::home_dir().unwrap().join(DEFAULT_WALLETS_VAULT_PATH);

    let directories = match list_dirs_in(path) {
        Ok(entries) => entries,
        Err(_) => {
            return 0;
        }
    };

    directories
        .into_iter()
        .map(num_of_wallets_in_vault)
        .max()
        .unwrap_or(1)
}

fn num_of_wallets_in_vault<T: AsRef<Path>>(dir: T) -> usize {
    extract_wallet_num_str(dir.as_ref())
        .unwrap()
        .parse()
        .unwrap()
}

fn extract_wallet_num_str(s: &Path) -> Option<&str> {
    s.file_name()?.to_str()?.split('_').next()
}

fn list_dirs_in(path: PathBuf) -> io::Result<Vec<PathBuf>> {
    let entries = std::fs::read_dir(path)?
        .into_iter()
        .collect::<Result<Vec<_>, _>>()?;

    Ok(entries
        .into_iter()
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .collect())
}
```

File: src/main.rs [hunk 4]
```diff
@@ -36,31 +46,153 @@ enum OutputFormat {
...
 
-fn print_wallet(wallet: Wallet, fmt: OutputFormat) {
-    let wallet = match fmt {
-        OutputFormat::Json => serde_json::to_string_pretty(&wallet).unwrap(),
-        OutputFormat::Toml => toml::to_string(&wallet).unwrap(),
+fn list_wallets() {
```
Comment (segfault-magnet) [inline src/main.rs:152]: For the functional lovers out there :D 
```rust
fn list_wallets() {
    let path = home::home_dir().unwrap().join(DEFAULT_WALLETS_VAULT_PATH);

    // list directories in the path
    let dirs = match list_dirs_in(path) {
        Ok(dirs) => dirs,
        Err(_) => {
            println!("No wallets found.");
            return;
        }
    };

    println!("Num.         Address.");

    for stringified_wallet in stringify_wallets(&dirs) {
        println!("{}", stringified_wallet);
    }
}

fn stringify_wallets(dirs: &[PathBuf]) -> Vec<String> {
    let mut abc = dirs
        .into_iter()
        .map(|dir| {
            let index = num_of_wallets_in_vault(dir);
            let address = extract_wallet_address(&dir).unwrap();

            let s = format!("[{}].         0x{}", index, address);
            (index, s)
        })
        .collect::<Vec<_>>();

    abc.sort_by_key(|(index, _)| *index);

    abc.into_iter().map(|(_, b)| b).collect()
}

fn extract_wallet_address(s: &Path) -> Option<&str> {
    s.file_name()?.to_str()?.split('_').last()
}

fn list_dirs_in(path: PathBuf) -> io::Result<Vec<PathBuf>> {
    let entries = std::fs::read_dir(path)?
        .into_iter()
        .collect::<Result<Vec<_>, _>>()?;

    Ok(entries
        .into_iter()
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .collect())
}
```

**Comment by iqdecay** [comment]
Appreciate the suggestions a lot, but this PR is more about putting "the basis" in, #11 is about refactoring.

**Review by Braqzen — COMMENTED** [review]
Couple questions

1. The `Command` enum takes 3 fields while the help message lists 4. Should the enum also contain a `help` field? When the `main` function is run it doesn't seem to match for `help` so I assume some magic happens somewhere and if `help` is seen then the program does not continue
2. Can someone explain to me the purpose of the `uuid` here?

**Comment by iqdecay** [comment]
- The `help` command is directly integrated inside the `clap` module.
- The UUID is, afaik, here to ensure that we generate wallets that have unique ids/names and avoid collisions.

**Review by iqdecay — APPROVED** [review]
still lgtm, give another approve please just want to move forward lol

**Review by Braqzen — APPROVED** [review]

**Pull request merged**